File: Linear System/Fortran/linear_system.f90
program linear_system

    !Undergraduate Student: Arturo Burgos
    !Professor: João Rodrigo Andrade
    !Federal University of Uberlândia - UFU, Fluid Mechanics Laboratory
    !MFLab, Block 5P, Uberlândia, MG, Brazil

    !Fourth exercise: Solving a Linear System --> ax = b


    implicit none
    
    integer :: i, j, k, n, ite
    real, DIMENSION(:), ALLOCATABLE :: mtx_b
    real, DIMENSION(:,:), ALLOCATABLE :: mtx_a
    DOUBLE PRECISION :: start, finish
    DOUBLE PRECISION :: erro, tol
    DOUBLE PRECISION, DIMENSION(:), ALLOCATABLE :: x_k, x_k1

    n = 400
    k = int(SQRT(real(n)))

    ALLOCATE (mtx_a(n,n))
    ALLOCATE (mtx_b(n))

    

    ! Set the A matrix

    do i = 1,n
        do j = 1,n
            if (i==j) then
                            
                mtx_a(i,j) = -4
            
            else if (i==j-3 .or. i==j+3) then

                mtx_a(i,j) = 1

            else if ((mod(i,3)/=0 .and. j==i+1).or.(mod(i,3)/=1 .and. j==i-1)) then
                
                mtx_a(i,j) = 1

            else

                mtx_a(i,j) = 0
            
            end if
        end do
    end do


    ! Set the B matrix

    do i = 1,k
        if (i<k) then

            mtx_b(i) = -50 
        
        else

            mtx_b(i) = -150
        
        endif
    end do

    do i = k+1,n-k
        if (mod(i,k)/=0) then

            mtx_b(i) = 0
        
        else

            mtx_b(i) = -100
        
        endif
    end do

    do i = n-k+1,n
        if (i<n) then

            mtx_b(i) = -50 
        
        else

            mtx_b(i) = -150
        
        endif
    end do
    
    ALLOCATE (x_k(n))
    ALLOCATE (x_k1(n))


    x_k(:) = 0
    x_k1(:) = 1


    erro = norma(x_k1,x_k)
    tol = 1e-9
    ite = 0


    call cpu_time(start)
    do while (erro > tol)

        do i = 1,n

            x_k1(i) = mtx_b(i)

            do j = 1,n
                if (j/=i) then

                    x_k1(i) = x_k1(i) - mtx_a(i,j)*x_k(j)

                endif
            end do

            x_k1(i) = x_k1(i)/mtx_a(i,i)

        end do

        erro = norma(x_k1,x_k)
        x_k = x_k1
        ite = ite + 1

    end do
    call cpu_time(finish)

    print *,'Elapsed time is:',finish - start
    print *, 'Solution: '
    print *, x_k1
    print *, 'Iterations: ',ite


    contains
    
    function norma(vec1,vec2) result(y)

        implicit none

        
        DOUBLE PRECISION, DIMENSION(:), ALLOCATABLE :: vec1, vec2
        DOUBLE PRECISION :: y

        
        y = maxval(ABS(vec1 - vec2))

    end function norma 


end program linear_system

File: Linear System/Python/linear_system.py
# Undergraduate Student: Arturo Burgos
# Professor: João Rodrigo Andrade
# Federal University of Uberlândia - UFU, Fluid Mechanics Laboratory - MFLab, Block 5P, Uberlândia, MG, Brazil


# Fourth exercise: Solving a Linear System --> ax = b

# Here I first set conditions

import numpy as np
from numpy import linalg as La
import matplotlib.pyplot as plt
import time
np.seterr(divide='ignore', invalid='ignore')

print('\n')

n = 324
K = np.sqrt(n)
k = int(K)


# Here I set the Matrix

a = np.zeros((n,n))


for i in range(n):
    for j in range(n):
        if i == j:
            a[i,j] = -4
        elif i == j-3 or i ==j+3:
            a[i,j] = 1
        elif ((i+1) % 3 != 0 and i == j-1) or ((i+1) % 3 != 1 and i == j+1): # (i+1) because in Python we start from 0
            a[i,j] = 1


print('The coefficient Matrix is:')
print(a)
print('\n')
        

b = np.zeros(n)


for i in range(k):
    if i < k-1: # (k-1) because in Python we start from 0
        b[i] = -50
    else:
        b[i] = -150

for i in range(k,n-k):
    if (i+1)%6 != 0: # (i+1) because in Python we start from 0
        b[i] = 0
    else:
        b[i] = -100

for i in range(n-k,n):
    if i < n-1: # (k-1) because in Python we start from 0
        b[i] = -50
    else:
        b[i] = -150

print('The result Matrix is:')
print(b)
print('\n')



def linearsystem(coeff,resul,size):

    # Initial x_k and x_k1 value

    x_k = np.zeros(size)
    x_k1 = np.ones(size)

    # Here I set the tolerance
    tolerance = 1e-9
    # Here I set the iterations
    ite = 0
  

    # Here I set the error based in the Infinite norm  
    erro = La.norm((x_k1 - x_k),np.inf)
    #erro = (x_k1 - x_k)/x_k1


    while (erro > tolerance): #
        for i in range(0,size):
        
            x_k1[i] = resul[i]

            for j in range(0,n):
                if j!=i:
                    x_k1[i] =  x_k1[i] - coeff[i,j]*x_k[j]


    
            x_k1[i] =  x_k1[i]/ coeff[i,i]


        #erro = (x_k1 - x_k)/x_k1
        erro = La.norm((x_k1 - x_k),np.inf)

        x_k = x_k1.copy()
        #x_k[:] = x_k1[:] # -> the same as above

        ite = ite + 1

    

    print('The number of iterations is: ')
    print(ite)
    print('\n')

    print('Note that now the error is not an array anymore, but is normalized :')
    print(erro)
    print('\n')

    return x_k1


t_initial = time.time()
res = linearsystem(a,b,n)
t_final = time.time()


print('The solution is:')
print(res)
print('\n')


print("\n\n--- %s seconds ---\n" % (t_final - t_initial))

# PLOT OF THE MATRIX

def samplemat(dims,bb):
    

    aa = np.zeros(dims)
    
    
    print(bb)

    aa = np.reshape(bb,(dims))

    
    return aa


# Display matrix
plt.matshow(samplemat((k, k),res))

plt.show()